File: Day264/nodeAtGivenDistanceInBT.java
//{ Driver Code Starts
//Initial Template for Java
package Day264;
import java.io.*;
import java.util.*;

class Node  
{ 
    int data; 
    Node left, right; 
   
    public Node(int d)  
    { 
        data = d; 
        left = right = null; 
    } 
}

class nodeAtGivenDistanceBT
{
    static Node buildTree(String str)
    {
        // Corner Case
        if(str.length() == 0 || str.equals('N'))
            return null;
        String[] s = str.split(" ");
        
        Node root = new Node(Integer.parseInt(s[0]));
        Queue <Node> q = new LinkedList<Node>();
        q.add(root);
        
        // Starting from the second element
        int i = 1;
        while(!q.isEmpty() && i < s.length)
        {
              // Get and remove the front of the queue
              Node currNode = q.remove();
        
              // Get the current node's value from the string
              String currVal = s[i];
        
              // If the left child is not null
              if(!currVal.equals("N")) 
              {
        
                  // Create the left child for the current node
                  currNode.left = new Node(Integer.parseInt(currVal));
        
                  // Push it to the queue
                  q.add(currNode.left);
              }
        
              // For the right child
              i++;
              if(i >= s.length)
                  break;
              currVal = s[i];
        
              // If the right child is not null
              if(!currVal.equals("N")) 
              {
        
                  // Create the right child for the current node
                  currNode.right = new Node(Integer.parseInt(currVal));
        
                  // Push it to the queue
                  q.add(currNode.right);
              }
              
              i++;
        }
    
        return root;
    }
    
    public static void main(String args[]) throws IOException
    {
    
       BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(br.readLine().trim());
        while(t>0)
        {
            String s = br.readLine().trim();
            Node root = buildTree(s);
            Solution T = new Solution();
            int target = Integer.parseInt(br.readLine().trim());
            int k = Integer.parseInt(br.readLine().trim());
            ArrayList<Integer> res = new ArrayList<Integer>();
            res = T.KDistanceNodes(root,target,k);
            for(int i = 0;i<res.size();i++)
                System.out.print(res.get(i) + " ");
            System.out.println();    
            t--;
        }
    }
}

// } Driver Code Ends


//User function Template for Java

// class Node  
// { 
//     int data; 
//     Node left, right;
// }

class Solution
{
    public static ArrayList<Integer> KDistanceNodes(Node root, int target , int k)
    {
        // return the sorted list of all nodes at k dist
        ArrayList<Integer> answer = new ArrayList<>();
        Map<Integer, Node> map = new HashMap<>();
        Queue<Node> queue = new LinkedList<>();
        queue.add(root);
        while (!queue.isEmpty()) {
            int size = queue.size();
            for (int i=0 ;i<size; i++) {
                Node temp = queue.poll();
                if (temp.left != null) {
                    queue.add(temp.left);
                    map.put(temp.left.data, temp);
                }
                if (temp.right != null) {
                    queue.add(temp.right);
                    map.put(temp.right.data, temp);
                }
            }
        }
        dfs(root, target, queue);
        Map<Integer, Integer> check = new HashMap<>();
        while (k >0 && !queue.isEmpty()) {
            int size = queue.size();
            for (int i=0; i<size; i++) {
                Node temp = queue.poll();
                check.put(temp.data, 1);
                if (temp.left!=null && !check.containsKey(temp.left.data)) {
                    queue.add(temp.left);
                }
                if (temp.right != null && !check.containsKey(temp.right.data)){
                    queue.add(temp.right);
                }
                if (map.containsKey(temp.data) && !check.containsKey(map.get(temp.data).data)){
                    queue.add(map.get(temp.data));
                }
            }
            k--;
        }
        while(!queue.isEmpty()) {
            answer.add(queue.poll().data);
        }
        Collections.sort(answer);
        return answer;
    }
    
    public static void dfs(Node root, int target , Queue<Node> queue) {
        if (root == null) {
            return;
        }
        if (root.data == target) {
            queue.add(root);
            return;
        }
        dfs(root.left, target, queue);
        dfs(root.right, target, queue);
        return;
    }
};

File: Day264/nodeAtGivenDistanceInBT.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type Node struct {
	data  int
	left  *Node
	right *Node
}

func buildTree(str string) *Node {
	if str == "" || str == "N" {
		return nil
	}
	s := strings.Split(str, " ")
	rootData, _ := strconv.Atoi(s[0])
	root := &Node{data: rootData}
	queue := []*Node{root}
	i := 1
	for i < len(s) && len(queue) > 0 {
		currNode := queue[0]
		queue = queue[1:]
		currVal := s[i]
		if currVal != "N" {
			leftData, _ := strconv.Atoi(currVal)
			currNode.left = &Node{data: leftData}
			queue = append(queue, currNode.left)
		}
		i++
		if i >= len(s) {
			break
		}
		currVal = s[i]
		if currVal != "N" {
			rightData, _ := strconv.Atoi(currVal)
			currNode.right = &Node{data: rightData}
			queue = append(queue, currNode.right)
		}
		i++
	}
	return root
}

func KDistanceNodes(root *Node, target, k int) []int {
	var answer []int
	var queue []*Node
	var check map[int]bool = make(map[int]bool)

	var dfs func(node *Node)
	dfs = func(node *Node) {
		if node == nil {
			return
		}
		if node.data == target {
			queue = append(queue, node)
			return
		}
		dfs(node.left)
		dfs(node.right)
	}

	dfs(root)

	for k > 0 && len(queue) > 0 {
		size := len(queue)
		for i := 0; i < size; i++ {
			temp := queue[0]
			queue = queue[1:]
			check[temp.data] = true
			if temp.left != nil && !check[temp.left.data] {
				queue = append(queue, temp.left)
			}
			if temp.right != nil && !check[temp.right.data] {
				queue = append(queue, temp.right)
			}
			if _, ok := check[temp.data]; ok {
				queue = append(queue, &Node{data: temp.data})
			}
		}
		k--
	}

	for len(queue) > 0 {
		answer = append(answer, queue[0].data)
		queue = queue[1:]
	}

	return answer
}

func main() {
	reader := bufio.NewReader(os.Stdin)
	tStr, _ := reader.ReadString('\n')
	t, _ := strconv.Atoi(strings.TrimSpace(tStr))
	fmt.Println("started")
	for t > 0 {
		s, _ := reader.ReadString('\n')
		targetStr, _ := reader.ReadString('\n')
		kStr, _ := reader.ReadString('\n')
		target, _ := strconv.Atoi(strings.TrimSpace(targetStr))
		k, _ := strconv.Atoi(strings.TrimSpace(kStr))
		root := buildTree(strings.TrimSpace(s))
		res := KDistanceNodes(root, target, k)
		for _, val := range res {
			fmt.Print(val, " ")
		}
		fmt.Println()
		t--
	}
}
